packages/web3-providers-ws/src/index.ts

```diff
@@ -0,0 +1,542 @@
+                        request
+                    );
+                    this.requestQueue.delete(key);
+                }
+            );
+        }
+
+        if (this.responseQueue.size > 0) {
+            this.responseQueue.forEach(
+                (request: RequestArguments, key: number) => {
+                    this.emit(
+                        WSStatus.Error, // TODO: Fancy errors,
+                        new Error(WSErrors.InvalidConnection),
+                        request
+                    );
+                    this.responseQueue.delete(key);
+                }
+            );
+        }
+
+        this.removeSocketListeners();
+    }
+
+    /**
+     * This function returns the status of Subscriptions boolean.
+     *
+     * @method supportsSubscriptions
+     *
+     * @returns {boolean}
+     */
+    supportsSubscriptions(): boolean {
+        return true;
+    }
+
+    /**
+     * This function creates w3cwebsocket client and adds Listeners.
+     *
+     * @method connect
+     *
+     * @returns {void}
+     */
+    connect(): void {
+        try {
+            this.webSocketConnection = new w3cwebsocket(
+                this.web3Client,
+                this.options && this.options.protocol
+                    ? this.options.protocol
+                    : undefined,
+                undefined,
+                this.options && this.options.headers
+                    ? this.options.headers
+                    : undefined,
+                this.options && this.options.requestOptions
+                    ? this.options.requestOptions
+                    : undefined,
+                this.options && this.options.clientConfig
+                    ? this.options.clientConfig
+                    : undefined
+            );
+
+            this.addSocketListeners();
+        } catch (error) {
+            throw Error(`Failed to create WebSocket client: ${error.message}`);
+        }
+    }
+
+    /**
+     *  This function closes the current connection with the given code and reason arguments
+     *
+     * @method disconnect
+     *
+     * @param {number} code
+     * @param {string} reason
+     *
+     * @returns {void}
+     */
+    async disconnect(code?: number, reason?: string): Promise<void> {
+        this.removeSocketListeners();
+
+        if (!this.webSocketConnection)
+            throw new Error(
+                'Cannot disconnect because of Invalid webSocketConnection'
+            );
+
+        this.webSocketConnection.close(code || 1000, reason);
+    }
+
+    /**
+     * This function is used to send request if provider is connected  else if the provider is connecting it will add request to the queue.
+     *
+     * @method request
+     *
+     * @param {RequestArguments} request
+     *
+     * @returns {RpcResponse} This function always returns undefined as result is emitted in events for web socket
+     */
+    async request(request: RequestArguments): Promise<void> {
```

Comment:
Right now `IWeb3Provider` specifies `request` as returning `Promise<RpcResponse>` via [Eip1193Provider](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-core-types/src/types.ts#L130), so `Promise<void>` is invalid regarding the interface. I think it's fine for now, since tsc isn't complaining when the project builds (even though it should be), but adjusting this needs to be addressed when fixing #4251 (I think we can just do `Promise<RpcResponse | void>`, but open to suggestions)

Reply:
It is already valid in this PR [`request: (args: RequestArguments) => Promise<any>`](https://github.com/ChainSafe/web3.js/blob/ed547ae2ab654dab7269ccd57b3fe6ee7f3edc2e/packages/web3-core-types/src/types.ts#L130)